python/ray/data/dataset.py

```diff
@@ -169,6 +174,8 @@ def __init__(
         plan: ExecutionPlan,
         epoch: int,
         lazy: bool,
+        *,
+        defer_execution: bool = False,
```

Comment:
I had hard time understand why there's a defer_execution when there is already a lazy parameter? Seems to muddy the execution semantics even more.